subsys/usb/device_next/class/usbd_cdc_ncm.c

```diff
@@ -0,0 +1,1282 @@
+			return false;
+		}
+		++ndx;
+	}
+
+	LOG_HEXDUMP_DBG(ntb->data, len, "NTB");
+
+	/* -> ntb contains a valid packet structure */
+	return true;
+}   /* _cdc_ncm_frame_ok */
+
+
+
+static int cdc_ncm_acl_out_cb(struct usbd_class_data *const c_data,
+		struct net_buf *const buf, const int err)
+/**
+ * Frame received from host!
+ */
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct cdc_ncm_eth_data *data = dev->data;
+	struct net_pkt *pkt;
+	union recv_ntb_t *ntb;
+	struct ndp16_datagram_t *ndp_datagram;
+
+	LOG_DBG("len %d err %d", buf->len, err);
+
+	if (err  ||  buf->len == 0) {
+		goto restart_out_transfer;
+	}
+
+	/* LOG_HEXDUMP_DBG(buf->data, buf->len, "ntb"); */
+
+	if (!_cdc_ncm_frame_ok(data, buf)) {
+		LOG_ERR("ill frame received from host");
+		goto restart_out_transfer;
+	}
+
+	ntb = (union recv_ntb_t *)buf->data;
+	ndp_datagram = (struct ndp16_datagram_t *)
+			(ntb->data + sys_le16_to_cpu(ntb->nth.wNdpIndex) + sizeof(struct ndp16_t));
+
+	uint16_t start = sys_le16_to_cpu(ndp_datagram[0].wDatagramIndex);
+	uint16_t len   = sys_le16_to_cpu(ndp_datagram[0].wDatagramLength);
+
+	pkt = net_pkt_rx_alloc_with_buffer(data->iface, len, AF_UNSPEC, 0, K_FOREVER);
+	if (pkt == NULL) {
+		LOG_ERR("No memory for net_pkt");
+		goto restart_out_transfer;
+	}
+
+	/*    LOG_HEXDUMP_DBG(buf->data + start, len, "frame"); */
+
+	if (net_pkt_write(pkt, ntb->data + start, len)) {
+		LOG_ERR("Unable to write into pkt");
+		net_pkt_unref(pkt);
+		goto restart_out_transfer;
+	}
+
+	LOG_DBG("Received packet len %zu", len);
+	if (net_recv_data(data->iface, pkt) < 0) {
+		LOG_ERR("Packet %p dropped by network stack", pkt);
+		net_pkt_unref(pkt);
+	}
+
+restart_out_transfer:
+	net_buf_unref(buf);
+	atomic_clear_bit(&data->state, CDC_NCM_OUT_ENGAGED);
+
+	return cdc_ncm_out_start(c_data);
+}   /* cdc_ncm_acl_out_cb */
+
+
+
+static int _usbd_cdc_ncm_send_notification(const struct device *dev,
+		const void *notification, uint16_t len)
+/**
+ * Send a notification to the host.
+ */
+{
+	struct cdc_ncm_eth_data *data = dev->data;
+	struct usbd_class_data *c_data = data->c_data;
+	struct net_buf *buf;
+	uint8_t ep;
+	int ret;
+
+	LOG_DBG("");
+
+	if (!atomic_test_bit(&data->state, CDC_NCM_CLASS_ENABLED)) {
+		LOG_INF("USB configuration is not enabled");
+		return 0;
+	}
+
+	if (atomic_test_bit(&data->state, CDC_NCM_CLASS_SUSPENDED)) {
+		LOG_INF("USB device is suspended (FIXME)");
+		return 0;
+	}
+
+	ep = cdc_ncm_get_int_in(c_data);
+	LOG_DBG("ep: 0x%02x", ep);
+	buf = usbd_ep_buf_alloc(c_data, ep, len);
+	if (buf == NULL) {
+		return -ENOMEM;
+	}
+
+	net_buf_add_mem(buf, notification, len);
+	ret = usbd_ep_enqueue(c_data, buf);
+	if (ret) {
+		LOG_ERR("Failed to enqueue net_buf for 0x%02x", ep);
+		net_buf_unref(buf);
+		return ret;
+	}
+
+	return 0;
+}   /* _usbd_cdc_ncm_send_notification */
+
+
+
+static void _usbd_cdc_ncm_notification_next_step(struct usbd_class_data *const c_data)
```

Comment:
renamed the three functions to z_*.  Idea behind naming is actually not the static keyword because almost everything is static in the driver.  It is to distinguish between API relevant functions and internal things

Reply:
I am also a bit puzzled about what to do I admit... more context: https://github.com/zephyrproject-rtos/zephyr/issues/58007

Reply:
Haha, so this recommends to revert to "_" as a prefix.  I will do this on request or on the next required change.

Reply:
ok... reverted back to "_"